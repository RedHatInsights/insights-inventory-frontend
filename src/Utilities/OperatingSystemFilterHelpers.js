import { OS_CHIP, operatingSystems } from './index';

export const toGroupSelectionValue = (osVersions) => osVersions.reduce((acc, version) => {
    const [majorVersion, minorVersion] = version.split('.');
    acc[`${majorVersion}.0`] = {
        ...(acc[`${majorVersion}.0`] || {}),
        [`${majorVersion}.${minorVersion}`]: true
    };
    return acc;
}, {});

export const groupOSVersions = (osVersions) => {
    const majorVersions = osVersions.filter(({ value }) => value.split('.')[1] === '0');

    return majorVersions.map((majorVersion) => ({
        ...majorVersion,
        groupSelectable: true,
        items: osVersions.filter(({ value }) => {
            const [majorOsVersion, minorOsVersion] = value.split('.');
            return majorOsVersion === majorVersion.value.split('.')[0] && minorOsVersion !== '0';
        }).sort(({ value }) => value.split('.')[1])
    }));
};

export const buildOSFilterConfig = (config = {}) => ({
    ...config,
    label: 'Operating System',
    type: 'group',
    filterValues: {
        selected: config.value,
        onChange: config.onChange,
        groups: groupOSVersions(operatingSystems).map((item) => ({
            ...item,
            type: 'checkbox',
            items: item.items?.map((subItem) => ({
                ...subItem,
                type: 'checkbox'
            }))
        }))
    }
});

export const buildOSChip = (operatingSystemValue) => {
    const value = Object.values(operatingSystemValue || {}).flatMap((majorOsVersion) => Object.keys(majorOsVersion));
    return (value?.length > 0 ? [{
        category: 'Operating System',
        type: OS_CHIP,
        chips: operatingSystems.filter(({ value: osValue }) => value.some((osFilterValue) => osFilterValue === osValue))
        .map(({ label, ...props }) => ({ name: label, ...props }))
    }] : []);};
