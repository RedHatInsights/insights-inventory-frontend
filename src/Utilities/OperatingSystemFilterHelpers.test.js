import { toGroupSelectionValue, buildOSFilterConfig, buildOSChip } from './OperatingSystemFilterHelpers';

describe('toGroupSelectionValue', () => {
    it('returns a select value for a group selection', () => {
        expect(toGroupSelectionValue(['7.0', '7.1', '7.2', '9.1', '8.5'])).toMatchSnapshot();
    });
});

describe('buildOSFilterConfig', function () {
    it('returns a filter configuration', () => {
        expect(buildOSFilterConfig()).toMatchSnapshot();
    });

    it('appends a given config', () => {
        expect(buildOSFilterConfig({
            filterValues: []
        })).toMatchSnapshot();
    });
});

describe('buildOSChip', function () {
    it('returnd chips', () => {
        expect(buildOSChip(['7.9'])).toMatchSnapshot();
    });
});
