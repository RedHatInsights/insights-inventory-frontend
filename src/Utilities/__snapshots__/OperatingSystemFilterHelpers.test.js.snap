// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildOSChip returnd chips 1`] = `
Array [
  Object {
    "category": "Operating System",
    "chips": Array [],
    "type": "operating_system",
  },
]
`;

exports[`buildOSFilterConfig appends a given config 1`] = `
Object {
  "filterValues": Object {
    "groups": Array [
      Object {
        "groupSelectable": true,
        "items": Array [],
        "label": "RHEL 9.0",
        "type": "checkbox",
        "value": "9.0",
      },
      Object {
        "groupSelectable": true,
        "items": Array [
          Object {
            "label": "RHEL 8.6",
            "type": "checkbox",
            "value": "8.6",
          },
          Object {
            "label": "RHEL 8.5",
            "type": "checkbox",
            "value": "8.5",
          },
          Object {
            "label": "RHEL 8.4",
            "type": "checkbox",
            "value": "8.4",
          },
          Object {
            "label": "RHEL 8.3",
            "type": "checkbox",
            "value": "8.3",
          },
          Object {
            "label": "RHEL 8.2",
            "type": "checkbox",
            "value": "8.2",
          },
          Object {
            "label": "RHEL 8.1",
            "type": "checkbox",
            "value": "8.1",
          },
        ],
        "label": "RHEL 8.0",
        "type": "checkbox",
        "value": "8.0",
      },
      Object {
        "groupSelectable": true,
        "items": Array [
          Object {
            "label": "RHEL 7.9",
            "type": "checkbox",
            "value": "7.9",
          },
          Object {
            "label": "RHEL 7.8",
            "type": "checkbox",
            "value": "7.8",
          },
          Object {
            "label": "RHEL 7.7",
            "type": "checkbox",
            "value": "7.7",
          },
          Object {
            "label": "RHEL 7.6",
            "type": "checkbox",
            "value": "7.6",
          },
          Object {
            "label": "RHEL 7.5",
            "type": "checkbox",
            "value": "7.5",
          },
          Object {
            "label": "RHEL 7.4",
            "type": "checkbox",
            "value": "7.4",
          },
          Object {
            "label": "RHEL 7.3",
            "type": "checkbox",
            "value": "7.3",
          },
          Object {
            "label": "RHEL 7.2",
            "type": "checkbox",
            "value": "7.2",
          },
          Object {
            "label": "RHEL 7.1",
            "type": "checkbox",
            "value": "7.1",
          },
        ],
        "label": "RHEL 7.0",
        "type": "checkbox",
        "value": "7.0",
      },
      Object {
        "groupSelectable": true,
        "items": Array [
          Object {
            "label": "RHEL 6.10",
            "type": "checkbox",
            "value": "6.10",
          },
          Object {
            "label": "RHEL 6.9",
            "type": "checkbox",
            "value": "6.9",
          },
          Object {
            "label": "RHEL 6.8",
            "type": "checkbox",
            "value": "6.8",
          },
          Object {
            "label": "RHEL 6.7",
            "type": "checkbox",
            "value": "6.7",
          },
          Object {
            "label": "RHEL 6.6",
            "type": "checkbox",
            "value": "6.6",
          },
          Object {
            "label": "RHEL 6.5",
            "type": "checkbox",
            "value": "6.5",
          },
          Object {
            "label": "RHEL 6.4",
            "type": "checkbox",
            "value": "6.4",
          },
          Object {
            "label": "RHEL 6.3",
            "type": "checkbox",
            "value": "6.3",
          },
          Object {
            "label": "RHEL 6.2",
            "type": "checkbox",
            "value": "6.2",
          },
          Object {
            "label": "RHEL 6.1",
            "type": "checkbox",
            "value": "6.1",
          },
        ],
        "label": "RHEL 6.0",
        "type": "checkbox",
        "value": "6.0",
      },
    ],
    "onChange": undefined,
    "selected": undefined,
  },
  "label": "Operating System",
  "type": "group",
}
`;

exports[`buildOSFilterConfig returns a filter configuration 1`] = `
Object {
  "filterValues": Object {
    "groups": Array [
      Object {
        "groupSelectable": true,
        "items": Array [],
        "label": "RHEL 9.0",
        "type": "checkbox",
        "value": "9.0",
      },
      Object {
        "groupSelectable": true,
        "items": Array [
          Object {
            "label": "RHEL 8.6",
            "type": "checkbox",
            "value": "8.6",
          },
          Object {
            "label": "RHEL 8.5",
            "type": "checkbox",
            "value": "8.5",
          },
          Object {
            "label": "RHEL 8.4",
            "type": "checkbox",
            "value": "8.4",
          },
          Object {
            "label": "RHEL 8.3",
            "type": "checkbox",
            "value": "8.3",
          },
          Object {
            "label": "RHEL 8.2",
            "type": "checkbox",
            "value": "8.2",
          },
          Object {
            "label": "RHEL 8.1",
            "type": "checkbox",
            "value": "8.1",
          },
        ],
        "label": "RHEL 8.0",
        "type": "checkbox",
        "value": "8.0",
      },
      Object {
        "groupSelectable": true,
        "items": Array [
          Object {
            "label": "RHEL 7.9",
            "type": "checkbox",
            "value": "7.9",
          },
          Object {
            "label": "RHEL 7.8",
            "type": "checkbox",
            "value": "7.8",
          },
          Object {
            "label": "RHEL 7.7",
            "type": "checkbox",
            "value": "7.7",
          },
          Object {
            "label": "RHEL 7.6",
            "type": "checkbox",
            "value": "7.6",
          },
          Object {
            "label": "RHEL 7.5",
            "type": "checkbox",
            "value": "7.5",
          },
          Object {
            "label": "RHEL 7.4",
            "type": "checkbox",
            "value": "7.4",
          },
          Object {
            "label": "RHEL 7.3",
            "type": "checkbox",
            "value": "7.3",
          },
          Object {
            "label": "RHEL 7.2",
            "type": "checkbox",
            "value": "7.2",
          },
          Object {
            "label": "RHEL 7.1",
            "type": "checkbox",
            "value": "7.1",
          },
        ],
        "label": "RHEL 7.0",
        "type": "checkbox",
        "value": "7.0",
      },
      Object {
        "groupSelectable": true,
        "items": Array [
          Object {
            "label": "RHEL 6.10",
            "type": "checkbox",
            "value": "6.10",
          },
          Object {
            "label": "RHEL 6.9",
            "type": "checkbox",
            "value": "6.9",
          },
          Object {
            "label": "RHEL 6.8",
            "type": "checkbox",
            "value": "6.8",
          },
          Object {
            "label": "RHEL 6.7",
            "type": "checkbox",
            "value": "6.7",
          },
          Object {
            "label": "RHEL 6.6",
            "type": "checkbox",
            "value": "6.6",
          },
          Object {
            "label": "RHEL 6.5",
            "type": "checkbox",
            "value": "6.5",
          },
          Object {
            "label": "RHEL 6.4",
            "type": "checkbox",
            "value": "6.4",
          },
          Object {
            "label": "RHEL 6.3",
            "type": "checkbox",
            "value": "6.3",
          },
          Object {
            "label": "RHEL 6.2",
            "type": "checkbox",
            "value": "6.2",
          },
          Object {
            "label": "RHEL 6.1",
            "type": "checkbox",
            "value": "6.1",
          },
        ],
        "label": "RHEL 6.0",
        "type": "checkbox",
        "value": "6.0",
      },
    ],
    "onChange": undefined,
    "selected": undefined,
  },
  "label": "Operating System",
  "type": "group",
}
`;

exports[`toGroupSelectionValue returns a select value for a group selection 1`] = `
Object {
  "7.0": Object {
    "7.0": true,
    "7.1": true,
    "7.2": true,
  },
  "8.0": Object {
    "8.5": true,
  },
  "9.0": Object {
    "9.1": true,
  },
}
`;
